internal/core/custom_resources/resources/alarms_lambda.go

```
@@ -0,0 +1,178 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-lambda-go/cfn"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+)
+
+const (
+	lambdaLoggedErrorAlarm    = "LambdaLoggedErrors"
+	lambdaLoggedWarnAlarm     = "LambdaLoggedWarns"
+	lambdaMemoryAlarm         = "LambdaHighMemory"
+	lambdaExecutionErrorAlarm = "LambdaExecutionErrors"
+	lambdaDurationAlarm       = "LambdaHighDuration"
+	lambdaThrottleAlarm       = "LambdaThrottles"
+)
+
+type LambdaAlarmProperties struct {
+	AlarmTopicArn string `validate:"required"`
+	FunctionName  string `validate:"required"`
+
+	FunctionMemoryMB   int `json:",string" validate:"omitempty,min=128,max=3008"`
+	FunctionTimeoutSec int `json:",string" validate:"min=0"`
+
+	// These are pointers because we have to distinguish 0 from not specified
+	LoggedErrorThreshold    *int `json:",string" validate:"omitempty,min=0"`
+	LoggedWarnThreshold     *int `json:",string" validate:"omitempty,min=0"`
+	ExecutionErrorThreshold *int `json:",string" validate:"omitempty,min=0"`
+	ThrottleThreshold       *int `json:",string" validate:"omitempty,min=0"`
+}
+
+func customLambdaAlarms(_ context.Context, event cfn.Event) (string, map[string]interface{}, error) {
+	switch event.RequestType {
+	case cfn.RequestCreate, cfn.RequestUpdate:
+		var props LambdaAlarmProperties
+		if err := parseProperties(event.ResourceProperties, &props); err != nil {
+			return "", nil, err
+		}
+
+		// Set defaults
+		// TODO - better parsing of properties with defaults
+		if props.FunctionMemoryMB == 0 {
```

I debated adding a `default` tag to the properties struct above, but then we'd have to use reflect to fill it in manually (unless there's a library we could use). I'm happy to do that, but it could get fairly complex if we wanted nested properties, lists, etc. Filling in the defaults manually like this is by far the simplest solution

Using reflect directly would have some other advantages over our current json marshal/unmarshal parsing - it'd be a bit faster (not very important) and it would allow us to warn about unused fields. Today, any mistyped custom resource property is simply ignored. This would presumably fail validation on your properties, but it'd be nice to have a warning

---

Why not just fail if not set? See my comment about using !Ref for these values for safety.

---

Since 90% of our Lambda functions use the default memory (128 MB), the idea was to use that as the default so we didn't have to specify the same value 20 times. This helps keep the template size smaller, for example

But per your earlier comment, we do need a way to keep these values aligned, in which case it would then be a required parameter, yes